cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/MapReduceTransformExecutorFactory.java

```diff
@@ -147,9 +152,14 @@ protected TrackedTransform getTransformation(String pluginType, String stageName
           stageMetrics, taskContext.getDataTracer(stageName));
       }
     }
-    return new TrackedTransform(KVTransformations.getKVTransformation(stageName, pluginType, isMapPhase,
-                                                                      getInitializedTransformation(stageName)),
-                                       stageMetrics, taskContext.getDataTracer(stageName));
+    Transformation transformation =
+      new TrackedTransform(KVTransformations.getKVTransformation(stageName, pluginType, isMapPhase,
+                                                                 getInitializedTransformation(stageName)),
+                           stageMetrics, taskContext.getDataTracer(stageName));
+    if (taskContext.getDataTracer(stageName).isEnabled() && BatchSource.PLUGIN_TYPE.equals(pluginType) && isMapPhase) {
+      return new LimitingTransform<>(transformation, numberOfRecordsPreview == 0 ? 100 : numberOfRecordsPreview);
```

Comment:
its better to do the defaulting at the place of origin instead of here. In other words, default when we read the config and not here.